src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java

```diff
@@ -22,47 +22,50 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
+package jdk.internal.event;
 
-package jdk.jfr.internal.instrument;
-
+import java.lang.reflect.Field;
 import java.util.concurrent.atomic.AtomicLong;
 
-import jdk.jfr.events.EventConfigurations;
-import jdk.jfr.events.ErrorThrownEvent;
-import jdk.jfr.events.ExceptionThrownEvent;
-import jdk.jfr.internal.event.EventConfiguration;
-
+/**
+ * Helper class for exception events.
+ */
 public final class ThrowableTracer {
 
     private static final AtomicLong numThrowables = new AtomicLong();
 
-    public static void traceError(Error e, String message) {
-        if (e instanceof OutOfMemoryError) {
+    public static void enable() throws NoSuchFieldException, IllegalAccessException {
+        Field field = Throwable.class.getDeclaredField("jfrTracing");
+        field.setAccessible(true);
+        field.setBoolean(null, true);
+    }
+
+    public static void traceError(Class<?> clazz, String message) {
+        if (OutOfMemoryError.class.isAssignableFrom(clazz)) {
             return;
         }
-        long timestamp = EventConfiguration.timestamp();
 
-        EventConfiguration eventConfiguration1 = EventConfigurations.ERROR_THROWN;
-        if (eventConfiguration1.isEnabled()) {
-            ErrorThrownEvent.commit(timestamp, message, e.getClass());
+        if (ErrorThrownEvent.enabled()) {
+            long timestamp = ErrorThrownEvent.timestamp();
+            ErrorThrownEvent.commit(timestamp, message, clazz);
         }
-        EventConfiguration eventConfiguration2 = EventConfigurations.EXCEPTION_THROWN;
-        if (eventConfiguration2.isEnabled()) {
-            ExceptionThrownEvent.commit(timestamp, message, e.getClass());
+        if (ExceptionThrownEvent.enabled()) {
+            long timestamp = ExceptionThrownEvent.timestamp();
+            ExceptionThrownEvent.commit(timestamp, message, clazz);
         }
         numThrowables.incrementAndGet();
     }
 
-    public static void traceThrowable(Throwable t, String message) {
-        EventConfiguration eventConfiguration = EventConfigurations.EXCEPTION_THROWN;
-        if (eventConfiguration.isEnabled()) {
-            long timestamp = EventConfiguration.timestamp();
-            ExceptionThrownEvent.commit(timestamp, message, t.getClass());
+    public static void traceThrowable(Class<?> clazz, String message) {
+        if (ExceptionThrownEvent.enabled()) {
+            long timestamp = ExceptionThrownEvent.timestamp();
+            ExceptionThrownEvent.commit(timestamp, message, clazz);
```

Comment:
Just a drive-by observation, but now you get the timestamp from the event it seems strange to pull it out and then pass it back to commit, instead of having a version of commit that automatically uses the internal timestamp.

Reply:
I agree, and I have looked into it, but I think it's better to do that refactorization seperately.

Reply:
Just for my own understanding: in this particular case the time stamp is meaningless because the duration is expected to be 0 (or close to it) since nothing happens between the time the timestamp is taken and the time the event is committed. Is that correct?

Reply:
Yes, the event has no duration.

Before https://bugs.openjdk.org/browse/JDK-8239508 we wrote timestamp and duration for all Java events. For exception events, we wrote duration = 0. Now we can differentiate between durational and instant events, which means bytecode can be generated to take the timestamp in the commit method.